OrbitVulkanLayer/TimerQueryPool.h

```diff
@@ -0,0 +1,146 @@
+    CHECK(result == VK_SUCCESS);
+
+    dispatch_table_->ResetQueryPoolEXT(device)(device, query_pool, 0, num_timer_query_slots_);
+
+    {
+      absl::WriterMutexLock lock(&mutex_);
+      device_to_query_pool_[device] = query_pool;
+      std::vector<internal::SlotState> slots{num_timer_query_slots_};
+      std::fill(slots.begin(), slots.end(), internal::SlotState::kReadyForQueryIssue);
+      device_to_query_slots_[device] = slots;
+      device_to_potential_next_free_index_[device] = 0;
+    }
+  }
+
+  /*
+   * Retrieves the query pool for a give device. Note, that the pool must be initialized using
+   * `InitializeTimerQueryPool` before.
+   */
+  [[nodiscard]] VkQueryPool GetQueryPool(VkDevice device) {
+    absl::ReaderMutexLock lock(&mutex_);
+    CHECK(device_to_query_pool_.contains(device));
+    return device_to_query_pool_.at(device);
+  }
+
+  /*
+   * Tries to find a free query slot in the device`s pool. It returns `false` if no slot was found
+   * (all slots are occupied) and true otherwise.
+   * In case it successfully found a slot, the index will be written to the given `allocated_index`.
+   *
+   * Note, that the pool pool must be initialized using `InitializeTimerQueryPool` before.
+   * See also `ResetQuerySlots` to make occupied slots available again.
+   */
+  [[nodiscard]] bool NextReadyQuerySlot(VkDevice device, uint32_t* allocated_index) {
+    absl::WriterMutexLock lock(&mutex_);
+    CHECK(device_to_potential_next_free_index_.contains(device));
+    CHECK(device_to_query_slots_.contains(device));
+    uint32_t potential_next_free_slot = device_to_potential_next_free_index_.at(device);
+    std::vector<internal::SlotState>& slots = device_to_query_slots_.at(device);
+    uint32_t current_slot = potential_next_free_slot;
+    do {
+      if (slots.at(current_slot) == internal::SlotState::kReadyForQueryIssue) {
+        device_to_potential_next_free_index_[device] = (current_slot + 1) % num_timer_query_slots_;
+        slots.at(current_slot) = internal::SlotState::kQueryPendingOnGPU;
+        *allocated_index = current_slot;
+        return true;
+      }
+      current_slot = (current_slot + 1) % num_timer_query_slots_;
+    } while (current_slot != potential_next_free_slot);
+
+    return false;
+  }
+
+  /*
+   * Resets an occupied slot to be ready for queries again.
+   * If `rollback_only` is set, it will not call to Vulkan to reset the content at that slot.
+   * This is useful, if the slot was retrieved, but the actual query was not yet submitted to
+   * Vulkan (e.g. if on resetting the command buffer).
+   *
+   * Note, that the pool pool must be initialized using `InitializeTimerQueryPool` before.
+   * Further, the given slots must be in the `kReadyForQueryIssue` state, i.e. must be a result
+   * of `NextReadyQuerySlot` and have not been reset yet.
+   */
+  void ResetQuerySlots(VkDevice device, const std::vector<uint32_t>& physical_slot_indices,
+                       bool rollback_only) {
```

Comment:
This `rollback_only` argument in the signature of a public method is a bit confusing, as this method basically has two distinct behaviors. Maybe what I would do is to have two distinct public methods that then delegate to an internal method that takes the bool (to avoid code duplication).

Reply:
Thought about the same, they actually used to be two different functions, that looked the same. I made it internal and provide two interfaces.